persistence/internal/providertest/eventstore/repository.go

```diff
@@ -196,30 +225,106 @@ func DeclareRepositoryTests(tc *common.TestContext) {
 					item5.Envelope,
 				)
 
-				q := eventstore.Query{
-					Filter: eventstore.NewFilter(
-						marshalfixtures.MessageAPortableName,
-					),
+				items := loadEventsBySource(
+					tc.Context,
+					repository,
+					"<aggregate>",
+					"<instance-a>",
+					"",
+				)
+
+				for i, item := range items {
+					expectItemToEqual(
+						item,
+						expectedA[i],
+						fmt.Sprintf("item at index #%d of slice", i),
+					)
 				}
 
-				var g sync.WaitGroup
+				items = loadEventsBySource(
+					tc.Context,
+					repository,
+					"<aggregate>",
+					"<instance-b>",
+					"",
+				)
 
-				fn := func() {
-					defer g.Done()
-					defer ginkgo.GinkgoRecover()
-					items := queryEvents(tc.Context, repository, q)
-					gomega.Expect(items).To(gomega.HaveLen(2))
+				for i, item := range items {
+					expectItemToEqual(
+						item,
+						expectedB[i],
+						fmt.Sprintf("item at index #%d of slice", i),
+					)
 				}
+			})
 
-				g.Add(3)
-				go fn()
-				go fn()
-				go fn()
-				g.Wait()
+			ginkgo.It("matches the events from the source only after the given message id", func() {
+				expectedA := []*eventstore.Item{item2}
+				expectedB := []*eventstore.Item{item3}
+
+				saveEvents(
+					tc.Context,
+					dataStore,
+					item0.Envelope,
+					item1.Envelope,
+					item2.Envelope,
+					item3.Envelope,
+					item4.Envelope,
+					item5.Envelope,
+				)
+
+				items := loadEventsBySource(
+					tc.Context,
+					repository,
+					"<aggregate>",
+					"<instance-a>",
+					item0.Envelope.MetaData.MessageId,
+				)
+
+				for i, item := range items {
+					expectItemToEqual(
+						item,
+						expectedA[i],
+						fmt.Sprintf("item at index #%d of slice", i),
+					)
+				}
+
+				items = loadEventsBySource(
+					tc.Context,
+					repository,
+					"<aggregate>",
+					"<instance-b>",
+					item1.Envelope.MetaData.MessageId,
+				)
+
+				for i, item := range items {
+					expectItemToEqual(
+						item,
+						expectedB[i],
+						fmt.Sprintf("item at index #%d of slice", i),
+					)
+				}
+			})
+
+			ginkgo.It("returns an error if the message with the given id is not found", func() {
+				saveEvents(
+					tc.Context,
+					dataStore,
+					item0.Envelope,
+					item1.Envelope,
+				)
+
+				_, err := repository.LoadEventsBySource(
+					tc.Context,
+					"<aggregate>",
+					"<instance-a>",
+					item2.Envelope.MetaData.MessageId,
+				)
+				gomega.Expect(err).Should(gomega.HaveOccurred())
```

Comment:
I've tried to avoid checking for non-specific errors in the public API. If there's an error condition that the provider HAS to produce, I've opted to make a custom error type part of the public API. 

There are a few benefits:

1) It makes the tests easier to understand, just because it's more explicit.
2) I've had real cases where such tests were returning some error, but it was for a totally different reason, and I had no idea.
3) It makes it that little bit clearer for implementors of new persistence providers (you :P)

So, that said, we'll need an `eventstore.UnknownMessageError` type, and to add this to the docs for `LoadEventsBySource()`.

Reply:
Cool, makes sense, will do.